sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -1192,13 +1218,34 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
           } else {
             s"$evNull =true;"
           }
+          val outOfDecimalCode = if (!ansiEnabled) {
+            s"$evNull = true;"
+          } else {
+            s"""throw new ArithmeticException("out of decimal type range: " + $c);"""
+          }
           code"""
+            java.math.BigDecimal $tmpBigDecimal = null;
             try {
-              Decimal $tmp = Decimal.apply(new java.math.BigDecimal($c.toString().trim()));
-              ${changePrecision(tmp, target, evPrim, evNull, canNullSafeCast)}
+              $tmpBigDecimal = new java.math.BigDecimal($c.toString().trim());
             } catch (java.lang.NumberFormatException e) {
               $handleException
             }
+
+            if ($tmpBigDecimal != null) {
```

Comment:
it's risky to duplicate the code in codegen. How about we create `Decimal.fromString` and `Decimal.fromStringANSI`?